File: bot/security.py
"""Security System - Anti-Scam Detection

Detects and prevents two main scammer types:
1. Wallet Scammers: Crypto wallet spam with currency symbols in name, ALL CAPS
2. Screenshot Spammers: 4+ crypto screenshots, cross-posting, gibberish text
"""
import datetime
import hashlib
import io
import json
import re
from typing import Dict, Optional

import discord
from PIL import Image

from .config import (
    logger,
    CATCHER_ROLE_ID,
    TRUSTED_USER_IDS,
    ADMIN_CHANNEL_IDS,
)

# ============================================================================
# MESSAGE TRACKING FOR CROSS-POSTING DETECTION
# ============================================================================

# Message tracking for cross-posting detection
# Structure: {user_id: [{'fingerprint': hash, 'channel_id': int, 'timestamp': float, 'message_id': int}, ...]}
user_recent_messages: Dict[int, list] = {}
MAX_TRACKED_MESSAGES_PER_USER = 50
CROSS_POST_WINDOW_SECONDS = 300  # 5 minutes

# Crypto scam keyword patterns (case-insensitive, with point values)
CRYPTO_SCAM_PATTERNS = {
    r"\bWALL?LET\b": 50,
    r"\b\d+\s*SOL\b": 50,
    r"\bDEAD\s+TOKENS?\b": 50,
    r"\bPAY\s+HIM\b": 50,
    r"\bPLENTY\s+TRANSACTIONS?\b": 40,
    r"\bEMPTY\s+WALLET\b": 40,
    r"\bCRYPTO\b": 20,
    r"\bDM\s+ME\b": 30,
    r"\bBUY\b.*\bWALLET\b": 40,
}

# ============================================================================
# MESSAGE FINGERPRINTING
# ============================================================================

def get_message_fingerprint(message: discord.Message) -> str:
    """Create a hash of message content + attachments for duplicate detection."""
    fingerprint = message.content.strip()
    for att in message.attachments:
        fingerprint += f"|{att.filename}|{att.size}"
    return hashlib.md5(fingerprint.encode()).hexdigest()


async def track_message(message: discord.Message):
    """Track message for cross-posting detection."""
    user_id = message.author.id
    fingerprint = get_message_fingerprint(message)
    timestamp = message.created_at.timestamp()

    if user_id not in user_recent_messages:
        user_recent_messages[user_id] = []

    user_recent_messages[user_id].append({
        "fingerprint": fingerprint,
        "channel_id": message.channel.id,
        "timestamp": timestamp,
        "message_id": message.id,
    })

    # Clean old messages (older than window)
    current_time = timestamp
    user_recent_messages[user_id] = [
        m for m in user_recent_messages[user_id]
        if current_time - m["timestamp"] < CROSS_POST_WINDOW_SECONDS
    ]

    # Limit tracking per user
    if len(user_recent_messages[user_id]) > MAX_TRACKED_MESSAGES_PER_USER:
        user_recent_messages[user_id] = user_recent_messages[user_id][-MAX_TRACKED_MESSAGES_PER_USER:]


async def check_cross_posting(message: discord.Message) -> int:
    """Check if user posted same content in multiple channels recently.

    Returns:
        Number of different channels where same message was posted

    """
    user_id = message.author.id
    if user_id not in user_recent_messages:
        return 0

    fingerprint = get_message_fingerprint(message)
    recent = user_recent_messages[user_id]

    # Find all messages with same fingerprint
    same_message_posts = [m for m in recent if m["fingerprint"] == fingerprint]

    # Count unique channels
    unique_channels = set(m["channel_id"] for m in same_message_posts)
    return len(unique_channels)

# ============================================================================
# SPAM & GIBBERISH DETECTION
# ============================================================================

def is_gibberish_or_spam(text: str, user_has_roles: bool = True, has_images: bool = False) -> bool:
    """Detect gibberish, spam, or suspicious text patterns.

    Args:
        text: Message content
        user_has_roles: Does user have roles beyond @everyone?
        has_images: Does message have image attachments?

    Returns:
        True if text appears to be gibberish/spam

    """
    text = text.strip()

    # Empty messages are only suspicious if there are no images
    if len(text) == 0:
        return not has_images

    # === ALLOW EMOTIONAL EXPRESSIONS FOR USERS WITH ROLES ===
    if user_has_roles:
        # "AAAA", "lmao", "omg", etc. are fine if user has roles
        # Pure repeated letter spam (like "AAAAAAaaaaaaaaAAAAA")
        unique_chars = set(text.replace(" ", "").lower())
        if len(unique_chars) <= 2:  # Only 1-2 unique letters
            return False  # Allow emotional spam for users with roles

    # === STRICT CHECKS FOR NO-ROLE USERS ===

    # Random letter string (like "tdnfaagoie")
    if text.isalpha() and " " not in text:
        if 5 <= len(text) <= 20:
            # Common okay words
            common_ok = [
                "hello", "hi", "thanks", "thank", "please", "welcome",
                "yes", "no", "okay", "ok", "sure", "nice", "good",
                "great", "awesome", "cool", "wow", "lol", "lmao",
                "rofl", "omg", "wtf", "brb", "afk", "gg", "gn",
            ]

            if text.lower() in common_ok:
                return False

            # If user has roles, be lenient
            if user_has_roles:
                return False

            # No roles + random string = gibberish
            return True

    return False

# ============================================================================
# SCAM SCORING
# ============================================================================

def calculate_wallet_scam_score(message: discord.Message) -> tuple[int, list[str]]:
    """Calculate scam likelihood score for wallet/crypto scammers.

    Returns:
        (score, reasons) where higher score = more likely scam
        100+ = instant ban
        75-99 = delete + alert admins
        50-74 = watchlist

    """
    score = 0
    reasons = []

    # Username checks
    name = message.author.display_name

    # Currency symbols in username (hoisting technique)
    if any(c in name for c in ["£", "€", "¥", "₿", "$", "₹", "₽"]):
        score += 20
        reasons.append("Currency symbols in username")

    # Hoisting characters (to appear at top of member list)
    if name and name[0] in ["!", "=", "#", "@", ".", "_", "-", "~"]:
        score += 20
        reasons.append("Hoisting character in username")

    # Check for auto-generated username patterns (e.g., word.word####_#####)
    if re.search(r"[a-z]+\.[a-z]+\d{2,4}_\d{4,}", name.lower()):
        score += 15
        reasons.append("Suspicious username pattern")

    # Caps spam in message
    if len(message.content) > 20:
        caps_count = sum(1 for c in message.content if c.isupper())
        caps_ratio = caps_count / len(message.content)
        if caps_ratio > 0.7:
            score += 30
            reasons.append(f"Caps spam ({caps_ratio*100:.0f}%)")

    # Check for crypto scam keywords
    content = message.content.upper()
    for pattern, points in CRYPTO_SCAM_PATTERNS.items():
        if re.search(pattern, content):
            score += points
            reasons.append(f"Keyword: {pattern}")

    # === ROLE CHECKS (FIXED) ===
    # Check if author is a real Member (with roles) before checking roles!
    if isinstance(message.author, discord.Member):

        # Only has CATCHER role (scammers exploit this to look legit)
        if CATCHER_ROLE_ID and len(message.author.roles) == 2:  # @everyone + CATCHER
            catcher_role = discord.utils.get(message.author.roles, id=CATCHER_ROLE_ID)
            if catcher_role:
                score += 30
                reasons.append("Only has CATCHER role")

        # No roles at all (even worse than just CATCHER)
        elif len(message.author.roles) == 1:  # Only @everyone
            score += 20
            reasons.append("No roles (only @everyone)")

    # If it is NOT a Member (e.g. Webhook/PluralKit/DM), we skip role checks entirely.

    # No profile picture (red flag, but not damning on its own)
    # Scammers often have Member + CATCHER + no profile pic, then post spam
    if message.author.avatar is None:
        score += 15
        reasons.append("No profile picture")

    # Note: We don't check account age - real scammers can be years old!

    return (score, reasons)

# ============================================================================
# IMAGE SAFETY VERIFICATION
# ============================================================================

def verify_image_safety(file_data: bytes, filename: str) -> tuple[bool, str]:
    """Check if file is actually an image, not malware.

    Returns:
        (is_safe, reason) where is_safe=False triggers instant ban

    """
    if len(file_data) < 4:
        return (False, "File too small")

    magic = file_data[:4]

    # Executables (INSTANT BAN)
    if magic[:2] == b"MZ":  # Windows .exe
        return (False, "Windows executable (.exe) disguised as image")

    if magic == b"\x7fELF":  # Linux binary
        return (False, "Linux binary (ELF) disguised as image")

    # Valid images
    if magic[:2] == b"\xff\xd8":  # JPEG
        return (True, "JPEG")
    if magic == b"\x89PNG":  # PNG
        return (True, "PNG")
    if magic[:4] == b"RIFF":  # WebP
        return (True, "WebP")
    if magic[:2] == b"BM":  # BMP
        return (True, "BMP")
    if magic[:3] == b"GIF":  # GIF
        return (True, "GIF")

    # Unknown format
    return (False, f"Unknown file format (magic: {magic.hex()})")

# ============================================================================
# BAN & DELETE ACTIONS
# ============================================================================

async def instant_ban(message: discord.Message, reason: str, details: list = None):
    """Ban user and delete all their recent messages."""
    logger.critical(f"🚨 INSTANT BAN: {message.author} ({message.author.id}) - {reason}")

    try:
        # Delete the message first
        await message.delete()

        # Delete all recent messages from this user (last 5 minutes)
        await delete_all_user_messages(message.author, message.channel.guild, minutes=5)

        # Ban the user
        ban_reason = f"Auto-ban: {reason}"
        if details:
            ban_reason += f" | {', '.join(details[:3])}"  # First 3 details

        await message.guild.ban(message.author, reason=ban_reason, delete_message_seconds=300)

        # Alert admins
        await alert_admins(message.guild, message.author, reason, details, action="BANNED")

    except discord.Forbidden:
        logger.error(f"❌ Missing permissions to ban {message.author}")
        await alert_admins(message.guild, message.author, reason, details, action="FAILED - Missing permissions")
    except Exception as e:
        logger.error(f"Error banning user: {e}")


async def delete_all_user_messages(user: discord.User, guild: discord.Guild, minutes: int = 5):
    """Delete all messages from a user in the guild from the last N minutes."""
    cutoff = datetime.datetime.now(datetime.timezone.utc) - datetime.timedelta(minutes=minutes)

    for channel in guild.text_channels:
        try:
            async for msg in channel.history(limit=100, after=cutoff):
                if msg.author.id == user.id:
                    await msg.delete()
                    logger.info(f"🗑️ Deleted message from {user} in {channel.name}")
        except discord.Forbidden:
            continue  # Skip channels bot can't access
        except Exception as e:
            logger.warning(f"Error deleting messages in {channel.name}: {e}")

# ============================================================================
# ADMIN ALERTS
# ============================================================================

async def alert_admins(guild: discord.Guild, user: discord.User, reason: str, details: list = None, action: str = "ALERT", bot=None):
    """Send alert to admin channels about security event (supports multiple channels).
    
    Args:
        guild: Discord guild where event occurred
        user: User involved in security event
        reason: Reason for alert
        details: Additional details list
        action: Action type (BANNED, COMPROMISED, DELETED, ALERT)
        bot: Bot instance (needed to access channels in other guilds)
    """
    if not ADMIN_CHANNEL_IDS:
        return  # No admin channels configured

    # Choose color based on severity
    color_map = {
        "BANNED": discord.Color.red(),           # Red = instant ban
        "COMPROMISED": discord.Color.gold(),     # Gold = possible hacked account
        "DELETED": discord.Color.orange(),       # Orange = suspicious but not banned
        "ALERT": discord.Color.yellow(),          # Yellow = low priority alert
    }
    embed_color = color_map.get(action, discord.Color.orange())

    embed = discord.Embed(
        title=f"🚨 Security {action}",
        description=f"**User:** {user.mention} (`{user.id}`)\n**Server:** {guild.name}\n**Reason:** {reason}",
        color=embed_color,
    )

    if details:
        embed.add_field(name="Details", value="\n".join(f"• {d}" for d in details[:10]))

    # Add special note for compromised accounts
    if action == "COMPROMISED":
        embed.add_field(
            name="⚠️ Action Required",
            value="This is a veteran account posting scam content. The account may be hacked. Consider:\n"
                  "• DM the user to verify their account security\n"
                  "• Temporarily mute them until they respond\n"
                  "• Do NOT ban unless confirmed malicious",
            inline=False,
        )

    embed.set_footer(text=f"User: {user}")
    if user.avatar:
        embed.set_thumbnail(url=user.avatar.url)

    # Send to all configured admin channels
    for channel_id in ADMIN_CHANNEL_IDS:
        try:
            # Try to get channel from the current guild first
            channel = guild.get_channel(channel_id)
            # If not in current guild, try to get from bot's accessible channels
            if not channel and bot:
                channel = bot.get_channel(channel_id)

            if channel:
                await channel.send(embed=embed)
            else:
                logger.warning(f"Admin channel {channel_id} not found or not accessible")
        except Exception as e:
            logger.error(f"Failed to send admin alert to channel {channel_id}: {e}")

# ============================================================================
# UTILITY FUNCTIONS
# ============================================================================

def reformat_json(string: str, indent: int = 2) -> Optional[str]:
    """Reformat JSON string with proper indentation.

    Args:
        string: JSON string to reformat
        indent: Indentation level

    Returns:
        Reformatted JSON or None if invalid

    """
    try:
        data = json.loads(string)
        return json.dumps(data, indent=indent)
    except json.JSONDecodeError:
        return None


def is_valid_image(image_data: bytes) -> bool:
    """Verify this is actually a valid image file."""
    if not image_data or len(image_data) < 100:
        return False

    try:
        with Image.open(io.BytesIO(image_data)) as img:
            width, height = img.size
            return width > 0 and height > 0
    except Exception:
        return False

File: src/lib/security.ts
import crypto from 'crypto';
import dns from 'dns';
import net from 'net';
import { Message, GuildMember, Guild, TextChannel, EmbedBuilder, Colors, PermissionFlagsBits } from 'discord.js';
import type { ResolvedModConfig } from './settings-types';
import { BLOCKED_IMAGE_DOMAINS } from './config';

// ── Cross-post tracking ───────────────────────────────────────────────────────

interface TrackedMessage { fingerprint: string; channelId: string; timestamp: number; bytes: number; isMedia: boolean; }
const userMessages = new Map<string, TrackedMessage[]>();
export const CROSS_POST_WINDOW = 300; // seconds; also the max retention, so velocity windows are clamped to it

function fingerprint(message: Message): string {
  let s = message.content.trim();
  for (const a of message.attachments.values()) s += `|${a.name}|${a.size}`;
  return crypto.createHash('md5').update(s).digest('hex');
}

// ── Media detection ───────────────────────────────────────────────────────────
// GIFs are frequently delivered as links from known hosts (Tenor/Giphy/Discord
// picker), not uploads. Scan message content synchronously — link embeds unfurl
// later via a separate MessageUpdate, too late for MessageCreate checks.
export function isGifLink(content: string, domains: string[]): boolean {
  if (!content || domains.length === 0) return false;
  const urls = content.match(/https?:\/\/\S+/gi);
  if (!urls) return false;
  for (const raw of urls) {
    let host: string;
    try {
      host = new URL(raw).hostname.toLowerCase();
    } catch {
      continue; // not a parseable URL — skip
    }
    // Exact host or a subdomain of a known host (proper parse, not substring match)
    if (domains.some(d => host === d.toLowerCase() || host.endsWith(`.${d.toLowerCase()}`))) {
      return true;
    }
  }
  return false;
}

export function isMediaMessage(message: Message, gifDomains: string[]): boolean {
  for (const a of message.attachments.values()) {
    const ct = a.contentType ?? '';
    // image/* and video/* — Discord often serves GIFs as video/mp4, and large
    // video uploads are a real spam payload, so both count toward velocity.
    if (ct.startsWith('image/') || ct.startsWith('video/')) return true;
  }
  return isGifLink(message.content ?? '', gifDomains);
}

export function trackMessage(message: Message, gifDomains: string[] = []): void {
  const uid = message.author.id;
  const now = Date.now() / 1000;
  const fp = fingerprint(message);
  let bytes = 0;
  for (const a of message.attachments.values()) bytes += a.size;
  const isMedia = isMediaMessage(message, gifDomains);
  const prev = (userMessages.get(uid) ?? []).filter(m => now - m.timestamp < CROSS_POST_WINDOW);
  prev.push({ fingerprint: fp, channelId: message.channelId, timestamp: now, bytes, isMedia });
  userMessages.set(uid, prev.slice(-50));
}

export function checkCrossPosting(message: Message): number {
  const uid = message.author.id;
  const fp = fingerprint(message);
  const recent = userMessages.get(uid) ?? [];
  const channels = new Set(recent.filter(m => m.fingerprint === fp).map(m => m.channelId));
  return channels.size;
}

// Two-track velocity over the same in-memory tracking, read over a tighter
// (configurable) window than CROSS_POST_WINDOW:
//   sameChannels  — distinct channels with the SAME fingerprint (identical repost)
//   mediaChannels — distinct channels carrying ANY media (catches different GIFs)
// windowSec is clamped by callers to CROSS_POST_WINDOW — entries older than that are pruned by trackMessage, so a larger window would silently undercount.
export function checkMediaVelocity(
  message: Message,
  windowSec: number,
): { sameChannels: number; mediaChannels: number; maxBytes: number } {
  const uid = message.author.id;
  const now = Date.now() / 1000;
  const fp = fingerprint(message);
  const recent = (userMessages.get(uid) ?? []).filter(m => now - m.timestamp < windowSec);
  const sameChannels = new Set(recent.filter(m => m.fingerprint === fp).map(m => m.channelId)).size;
  const mediaMsgs = recent.filter(m => m.isMedia);
  const mediaChannels = new Set(mediaMsgs.map(m => m.channelId)).size;
  const maxBytes = mediaMsgs.reduce((mx, m) => Math.max(mx, m.bytes), 0);
  return { sameChannels, mediaChannels, maxBytes };
}

// ── Gibberish / spam detection ────────────────────────────────────────────────

const COMMON_OK = new Set([
  'hello','hi','thanks','thank','please','welcome','yes','no','okay','ok',
  'sure','nice','good','great','awesome','cool','wow','lol','lmao','rofl',
  'omg','wtf','brb','afk','gg','gn',
]);

export function isGibberish(text: string, userHasRoles: boolean, hasImages: boolean): boolean {
  text = text.trim();
  if (!text) return !hasImages;

  if (userHasRoles) {
    const unique = new Set(text.replace(/\s/g, '').toLowerCase().split(''));
    if (unique.size <= 2) return false;
  }

  if (/^[a-zA-Z]+$/.test(text) && !text.includes(' ') && text.length >= 5 && text.length <= 20) {
    if (COMMON_OK.has(text.toLowerCase())) return false;
    if (userHasRoles) return false;
    return true;
  }

  return false;
}

// ── Wallet scam scoring ───────────────────────────────────────────────────────

const SCAM_PATTERNS: [RegExp, number][] = [
  [/\bWALL?LET\b/i, 50],
  [/\b\d+\s*SOL\b/i, 50],
  [/\bDEAD\s+TOKENS?\b/i, 50],
  [/\bPAY\s+HIM\b/i, 50],
  [/\bPLENTY\s+TRANSACTIONS?\b/i, 40],
  [/\bEMPTY\s+WALLET\b/i, 40],
  [/\bCRYPTO\b/i, 20],
  [/\bDM\s+ME\b/i, 30],
  [/\bBUY\b.*\bWALLET\b/i, 40],
];

export function calculateScamScore(message: Message, cfg: ResolvedModConfig): [number, string[]] {
  let score = 0;
  const reasons: string[] = [];
  const name = (message.member?.displayName ?? message.author.username);

  if (/[£€¥₿$₹₽]/.test(name)) { score += 20; reasons.push('Currency symbols in username'); }
  if (name && /^[!=@#._\-~]/.test(name)) { score += 20; reasons.push('Hoisting character in username'); }
  if (/[a-z]+\.[a-z]+\d{2,4}_\d{4,}/.test(name.toLowerCase())) { score += 15; reasons.push('Auto-generated username pattern'); }

  if (message.content.length > 20) {
    const caps = [...message.content].filter(c => c >= 'A' && c <= 'Z').length;
    const ratio = caps / message.content.length;
    if (ratio > 0.7) { score += 30; reasons.push(`Caps spam (${Math.round(ratio * 100)}%)`); }
  }

  for (const [pattern, pts] of SCAM_PATTERNS) {
    if (pattern.test(message.content)) { score += pts; reasons.push(`Keyword match: ${pattern.source}`); }
  }

  const member = message.member;
  if (member) {
    const roles = member.roles.cache;
    if (cfg.catcherRoleId && roles.size === 2 && roles.has(cfg.catcherRoleId)) {
      score += 30; reasons.push('Only has CATCHER role');
    } else if (roles.size === 1) {
      score += 20; reasons.push('No roles (only @everyone)');
    }
  }

  if (!message.author.avatar) { score += 15; reasons.push('No profile picture'); }

  return [score, reasons];
}

// ── Magic bytes check ─────────────────────────────────────────────────────────

// Detects ONLY a binary executable disguised as an image (MZ / ELF) — the genuine
// attack the magic-bytes check exists to stop. Returns a reason string when the
// bytes are a known executable, otherwise null. Crucially, "this isn't a format I
// recognise" (JSON error pages, SVG, expired-CDN responses) is NOT malicious and
// returns null — banning on unverifiable content false-bans real users.
export function detectDisguisedExecutable(data: Buffer): string | null {
  if (data.length < 2) return null;
  if (data[0] === 0x4D && data[1] === 0x5A) return 'Windows executable disguised as image';
  if (data[0] === 0x7F && data[1] === 0x45) return 'Linux ELF binary disguised as image';
  return null;
}

export function verifyImageSafety(data: Buffer, filename: string): [boolean, string] {
  if (data.length < 4) return [false, 'File too small'];
  const exe = detectDisguisedExecutable(data);
  if (exe) return [false, exe];
  const magic = data.subarray(0, 4);
  if (magic[0] === 0xFF && magic[1] === 0xD8) return [true, 'JPEG'];
  if (magic.toString('ascii', 1, 4) === 'PNG') return [true, 'PNG'];
  if (data.subarray(0, 4).toString('ascii') === 'RIFF') return [true, 'WebP'];
  if (magic[0] === 0x42 && magic[1] === 0x4D) return [true, 'BMP'];
  if (data.subarray(0, 3).toString('ascii') === 'GIF') return [true, 'GIF'];
  return [false, `Unknown format (magic: ${magic.toString('hex')})`];
}

// ── Admin alert ───────────────────────────────────────────────────────────────

const ACTION_COLORS: Record<string, number> = {
  BANNED: Colors.Red,
  FAILED: Colors.Red,
  COMPROMISED: Colors.Yellow,
  DELETED: Colors.Orange,
  ALERT: Colors.Yellow,
};

export async function alertAdmins(
  guild: Guild,
  member: GuildMember | { id: string; displayName: string; avatarURL?: () => string | null },
  reason: string,
  details: string[],
  action: string,
  cfg: ResolvedModConfig,
): Promise<void> {
  if (!cfg.alertChannelIds.size) return;

  const embed = new EmbedBuilder()
    .setColor(ACTION_COLORS[action] ?? Colors.Orange)
    .setTitle(`🚨 Security ${action}`)
    .setDescription(`**User:** <@${member.id}> (\`${member.id}\`)\n**Server:** ${guild.name}\n**Reason:** ${reason}`)
    .setFooter({ text: String(member.id) });

  if (details.length) embed.addFields({ name: 'Details', value: details.slice(0, 10).map(d => `• ${d}`).join('\n') });

  const avatar = typeof member.avatarURL === 'function' ? member.avatarURL() : null;
  if (avatar) embed.setThumbnail(avatar);

  for (const channelId of cfg.alertChannelIds) {
    const channel = guild.channels.cache.get(channelId) as TextChannel | undefined;
    if (channel) await channel.send({ embeds: [embed] }).catch(() => null);
  }
}

// ── Instant ban ───────────────────────────────────────────────────────────────

export async function instantBan(message: Message, reason: string, cfg: ResolvedModConfig, details: string[] = []): Promise<void> {
  console.error(`🚨 BAN: ${message.author.tag} (${message.author.id}) — ${reason}`);
  if (!message.guild) return;

  const me = message.guild.members.me;
  if (!me || !me.permissions.has(PermissionFlagsBits.BanMembers)) {
    await alertAdmins(message.guild, message.member ?? message.author as any,
      reason, [...details, 'Bot missing BAN_MEMBERS permission'], 'FAILED', cfg);
    return;
  }

  try {
    await message.delete().catch(() => null);
    await message.guild.members.ban(message.author.id, {
      reason: `Auto-ban: ${reason} | ${details.slice(0, 3).join(', ')}`,
      deleteMessageSeconds: 300,
    });
    await alertAdmins(message.guild, message.member ?? message.author as any, reason, details, 'BANNED', cfg);
  } catch (e) {
    console.error('Ban failed:', e);
    if (message.guild) await alertAdmins(message.guild, message.member ?? message.author as any, reason, details, 'FAILED', cfg);
  }
}

// ── Algo speak detection ──────────────────────────────────────────────────────
// Targets the specific obfuscation patterns used by illegal content bots:
// zero-width characters, zalgo stacking, and homoglyph script mixing.
// NOT intended as a general Unicode filter — legitimate multilingual text scores 0.

const ZERO_WIDTH = /[​‌‍﻿⁠­]/g;
const COMBINING  = /[̀-ͯ᷀-᷿⃐-⃿︠-︯]/g;
// Cyrillic/Greek lookalikes for Latin letters (homoglyphs commonly used to evade filters)
const HOMOGLYPH_CYRILLIC = /[аеіорсухц]/g;

export function algoSpeakScore(text: string): number {
  if (!text || text.length < 5) return 0;
  let score = 0;

  // Zero-width / invisible characters — almost never legitimate in Discord chat
  const zwMatches = text.match(ZERO_WIDTH) ?? [];
  if (zwMatches.length >= 1) score += 40;
  if (zwMatches.length >= 3) score += 30; // cumulative — heavily stacked = near-certain evasion

  // Zalgo / combining diacritic stacking
  const combMatches = text.match(COMBINING) ?? [];
  const combRatio = combMatches.length / text.length;
  if (combRatio > 0.1) score += 30;
  if (combRatio > 0.3) score += 30;

  // Cyrillic homoglyphs mixed into otherwise Latin text
  const cyrMatches = text.match(HOMOGLYPH_CYRILLIC) ?? [];
  const latinCount = (text.match(/[a-zA-Z]/g) ?? []).length;
  if (cyrMatches.length > 0 && latinCount > 0) {
    const mixRatio = cyrMatches.length / (cyrMatches.length + latinCount);
    // Low ratio = deliberate substitution of a few chars (classic homoglyph attack)
    if (mixRatio > 0 && mixRatio < 0.3) score += 35;
  }

  return score;
}

// ── SSRF guard ────────────────────────────────────────────────────────────────

function isPrivateIP(ip: string): boolean {
  const n = ip.split('.').map(Number);
  if (n.length !== 4) return false;
  // 127.0.0.0/8
  if (n[0] === 127) return true;
  // 10.0.0.0/8
  if (n[0] === 10) return true;
  // 169.254.0.0/16
  if (n[0] === 169 && n[1] === 254) return true;
  // 172.16.0.0/12
  if (n[0] === 172 && n[1] >= 16 && n[1] <= 31) return true;
  // 192.168.0.0/16
  if (n[0] === 192 && n[1] === 168) return true;
  return false;
}

async function resolveAndCheckURL(rawUrl: string): Promise<string | null> {
  let host: string;
  try {
    host = new URL(rawUrl).hostname;
  } catch {
    return 'Invalid URL in embed';
  }

  const lowerHost = host.toLowerCase();

  // Blocked domain check
  for (const blocked of BLOCKED_IMAGE_DOMAINS) {
    if (lowerHost === blocked || lowerHost.endsWith(`.${blocked}`)) {
      return `Blocked domain: ${host}`;
    }
  }

  // DNS resolution — skip fetch if any A record points to a private/loopback IP
  try {
    const addresses = await dns.promises.resolve4(host);
    for (const addr of addresses) {
      if (isPrivateIP(addr)) {
        return `Blocked private IP (${addr}) for host: ${host}`;
      }
    }
  } catch {
    // DNS failure — allow the fetch (rate-limited below, might just be a transient error)
  }

  return null; // OK to proceed
}

// ── Embed URL magic bytes check ───────────────────────────────────────────────

export async function checkEmbedImages(message: Message): Promise<string | null> {
  for (const embed of message.embeds) {
    const url = embed.image?.url ?? embed.thumbnail?.url;
    if (!url) continue;

    const ssrfReason = await resolveAndCheckURL(url);
    if (ssrfReason) return ssrfReason;

    try {
      const res = await fetch(url, { signal: AbortSignal.timeout(5000) });
      const buf = Buffer.from(await res.arrayBuffer());
      // Only ban on a genuinely malicious payload (an executable disguised as an
      // image). Embed image URLs routinely resolve to non-image content — expired
      // Discord CDN links return JSON, link previews can return SVG/HTML — and that
      // is not an attack. Treating "unverifiable" as "malicious" false-bans bots
      // (e.g. Carlbot log embeds) and real users posting expired links.
      const exeReason = detectDisguisedExecutable(buf);
      if (exeReason) return exeReason;
    } catch { /* network error — skip */ }
  }
  return null;
}

// ── Mention spam detection ───────────────────────────────────────────────────

export function checkMentionSpam(message: Message): [number, string[]] {
  let score = 0;
  const reasons: string[] = [];

  if (message.mentions.everyone) {
    score += 50;
    reasons.push('@everyone or @here mention');
  }

  const userMentions = message.mentions.users.size;
  // Bot accounts used for mass-mention attacks target many distinct users at once;
  // 5+ unique user pings in a single message is well outside normal behaviour.
  if (userMentions >= 5) {
    score += Math.min(30 + userMentions * 2, 80);
    reasons.push(`Mass mention (${userMentions} users)`);
  }

  return [score, reasons];
}

// ── Bypass checks ─────────────────────────────────────────────────────────────

// Honeypot/catcher role: fires on mere presence (even if the member also holds a
// verified role), unlike the score nudge in calculateScamScore which only counts
// it when it is the only role.
export function hasHoneypotRole(message: Message, cfg: ResolvedModConfig): boolean {
  if (!cfg.catcherRoleId) return false;
  return message.member?.roles?.cache?.has(cfg.catcherRoleId) ?? false;
}

export function isTrusted(message: Message, cfg: ResolvedModConfig): boolean {
  if (cfg.trustedUserIds.has(message.author.id)) return true;
  if (message.guild && message.author.id === message.guild.ownerId) return true;
  if (cfg.trustedRoleIds.size) {
    // message.member is null for webhook/interaction bot messages (e.g. Carlbot),
    // so a trusted role would never match. Fall back to the guild's member cache
    // (cache-only — no fetch, keeps this synchronous) so a trusted role can still
    // exempt a bot that is already a known guild member.
    const roles =
      message.member?.roles?.cache ??
      message.guild?.members?.cache?.get(message.author.id)?.roles?.cache;
    if (roles) {
      for (const roleId of cfg.trustedRoleIds) if (roles.has(roleId)) return true;
    }
  }
  return false;
}
